pkg/kademlia/lookup.go

```diff
@@ -4,89 +4,135 @@
-		neighbors := lookup.FetchNeighbors(ctx, next)
-		for _, neighbor := range neighbors {
-			if !lookup.contacted[neighbor.GetId()] {
-				uncontactedNeighbors = append(uncontactedNeighbors, neighbor)
-			}
-		}
-		lookup.queue.Insert(lookup.target, uncontactedNeighbors)
+		cond:  &sync.Cond{L: &sync.Mutex{}},
+		queue: queue,
 
-		for lookup.queue.Len() > lookup.limit {
-			lookup.queue.Closest()
-		}
+		mu:        &sync.Mutex{},
+		contacted: map[string]int{},
 	}
-	return nil
 }
 
-func (lookup *sequentialLookup) FetchNeighbors(ctx context.Context, node *pb.Node) []*pb.Node {
-	if node.GetAddress() == nil {
-		return nil
-	}
-	lookup.contacted[node.GetId()] = true
-
-	start := time.Now()
-	neighbors, err := lookup.client.Lookup(ctx, *node, pb.Node{Id: lookup.target.String()})
-	if err != nil {
-		// TODO(coyle): I think we might want to do another look up on this node or update something
-		// but for now let's just log and ignore.
-		log.Printf("Error occurred during lookup for %s on %s :: error = %s", lookup.target.String(), node.GetId(), err.Error())
-		return []*pb.Node{}
-	}
+func (lookup *concurrentLookup) Run(ctx context.Context) error {
+	wg := sync.WaitGroup{}
+
+	// protected by `lookup.cond.L`
+	working := 0
+	allDone := false
+
+	wg.Add(lookup.opts.concurrency)
+	for i := 0; i < lookup.opts.concurrency; i++ {
+		go func() {
+			defer wg.Done()
+			for {
+				var (
+					next     *pb.Node
+					priority big.Int
+				)
+
+				lookup.cond.L.Lock()
+				for {
+					// everything is done, this routine can return
+					if allDone {
+						lookup.cond.L.Unlock()
+						return
+					}
+
+					next, priority = lookup.queue.Closest()
+					if !lookup.opts.bootstrap && priority.Cmp(&big.Int{}) == 0 {
+						allDone = true
+						lookup.cond.L.Unlock()
+						return // closest node is the target and is already in routing table (i.e. no lookup required)
+					}
 
-	latency := time.Since(start)
-	if latency > lookup.slowestResponse {
-		lookup.slowestResponse = latency
+					if next != nil {
+						working++
+						break
+					}
+
+					// no work, wait until some other routine inserts into the queue
+					lookup.cond.Wait()
+				}
+				lookup.cond.L.Unlock()
+
+				nextID := next.GetId()
+				lookup.mu.Lock()
+				tries := lookup.contacted[nextID]
+				lookup.contacted[nextID]++
+				lookup.mu.Unlock()
+
+				var uncontacted []*pb.Node
+				neighbors, err := lookup.client.Lookup(ctx, *next, pb.Node{Id: lookup.target.String()})
+				if err != nil {
+					if tries < lookup.opts.retries {
+						uncontacted = append(uncontacted, next)
+					} else {
+						log.Printf("Error occurred during lookup for %s on %s :: error = %s", lookup.target.String(), next.GetId(), err.Error())
+					}
+				}
+
+				for _, neighbor := range neighbors {
+					lookup.mu.Lock()
+					contacted := lookup.contacted[neighbor.GetId()] > 0
+					lookup.mu.Unlock()
+
+					if !contacted {
+						uncontacted = append(uncontacted, neighbor)
```

Comment:
You may end up adding a single neighbor multiple times into uncontacted. Add the tracking inside xorqueue, rather than handling it here. This should also remove the need to have a separate mutex for the uncontacted part.

For retry create a separate func on the queue `Readd(item node, limit int)`.

Reply:
can also make it a map to avoid duplicates